samples/s03-singleAABB/hostCode.cpp

```diff
@@ -0,0 +1,267 @@
+
+  // ##################################################################
+  // build the pipeline and shader binding table
+  // ##################################################################
+  gprtBuildPipeline(context);
+  gprtBuildShaderBindingTable(context);
+
+  // ##################################################################
+  // now that everything is ready: launch it ....
+  // ##################################################################
+
+  LOG("launching ...");
+
+  bool firstFrame = true;
+  double xpos = 0.f, ypos = 0.f;
+  double lastxpos, lastypos;
+  do 
+  {
+    float speed = .001f;
+    lastxpos = xpos;
+    lastypos = ypos;
+    gprtGetCursorPos(context, &xpos, &ypos);
+    if (firstFrame) {
+      lastxpos = xpos;
+      lastypos = ypos;
+    }
+    int state = gprtGetMouseButton(context, GPRT_MOUSE_BUTTON_LEFT);
+
+    // If we click the mouse, we should rotate the camera
+    // Here, we implement some simple camera controls
+    if (state == GPRT_PRESS || firstFrame)
+    {
+      firstFrame = false;
+      float4 position = {lookFrom.x, lookFrom.y, lookFrom.z, 1.f};
+      float4 pivot = {lookAt.x, lookAt.y, lookAt.z, 1.0};
+      #ifndef M_PI
+      #define M_PI 3.1415926f
+      #endif
+
+      // step 1 : Calculate the amount of rotation given the mouse movement.
+      float deltaAngleX = (2 * M_PI / fbSize.x);
+      float deltaAngleY = (M_PI / fbSize.y);
+      float xAngle = (lastxpos - xpos) * deltaAngleX;
+      float yAngle = (lastypos - ypos) * deltaAngleY;
+
+      // step 2: Rotate the camera around the pivot point on the first axis.
+      float4x4 rotationMatrixX = rotation_matrix(rotation_quat(lookUp, xAngle));
+      position = (mul(rotationMatrixX, (position - pivot))) + pivot;
+
+      // step 3: Rotate the camera around the pivot point on the second axis.
+      float3 lookRight = cross(lookUp, normalize(pivot - position).xyz());
+      float4x4 rotationMatrixY = rotation_matrix(rotation_quat(lookRight, yAngle));
+      lookFrom = ((mul(rotationMatrixY, (position - pivot))) + pivot).xyz();
+
+      // ----------- compute variable values  ------------------
+      float3 camera_pos = lookFrom;
+      float3 camera_d00
+        = normalize(lookAt-lookFrom);
+      float aspect = float(fbSize.x) / float(fbSize.y);
+      float3 camera_ddu
+        = cosFovy * aspect * normalize(cross(camera_d00,lookUp));
+      float3 camera_ddv
+        = cosFovy * normalize(cross(camera_ddu,camera_d00));
+      camera_d00 -= 0.5f * camera_ddu;
+      camera_d00 -= 0.5f * camera_ddv;
+
+      // ----------- set variables  ----------------------------
+      gprtRayGenSet3fv    (rayGen,"camera.pos",   (float*)&camera_pos);
+      gprtRayGenSet3fv    (rayGen,"camera.dir_00",(float*)&camera_d00);
+      gprtRayGenSet3fv    (rayGen,"camera.dir_du",(float*)&camera_ddu);
+      gprtRayGenSet3fv    (rayGen,"camera.dir_dv",(float*)&camera_ddv);
+
+      gprtBuildShaderBindingTable(context, GPRT_SBT_RAYGEN);
```

Comment:
Let's add a comment here on why this is necessary before we ray trace if they view has changed.